google-cloud-spanner/src/main/java/com/google/cloud/spanner/AsyncResultSetImpl.java

```diff
@@ -458,16 +488,24 @@ public ApiFuture<Void> setCallback(Executor exec, ReadyCallback cb) {
           this.state == State.INITIALIZED, "callback may not be set multiple times");
 
       // Start to fetch data and buffer these.
-      this.result =
-          new ListenableFutureToApiFuture<>(this.service.submit(new ProduceRowsCallable()));
+      this.result = SettableApiFuture.create();
+      this.state = State.STREAMING_INITIALIZED;
+      this.service.execute(new InitiateStreamingRunnable());
       this.executor = MoreExecutors.newSequentialExecutor(Preconditions.checkNotNull(exec));
       this.callback = Preconditions.checkNotNull(cb);
-      this.state = State.RUNNING;
       pausedLatch.countDown();
       return result;
     }
   }
 
+  private void initiateProduceRows() {
+    this.service.execute(new ProduceRowsRunnable());
```

Comment:
I think that you should move this line to the end of this method to be absolutely safe against weird race conditions. This method is called both from `InitiateStreamingRunnable` and from `onStreamMessage`. That theoretically means that:
1. `InitiateStreamingRunnable` calls this method and then triggers `ProduceRowsRunnable` here.
2. Before this thread continues with the next line, `ProduceRowsRunnable` has already been started and receives a `PartialResultSet`, and then calls `initiateProduceRows()`. (This is highly unlikely, due to the fact that this depends on a network call, but I still think it would be safer to change the status flags before starting `ProduceRowsRunnable`)

Reply:
Agreed. its very rate. but it should be safe to do it at last.